File: src/merge_video.cpp
#include <opencv2/opencv.hpp>
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
#include <filesystem>
#include <windows.h>
#include <shlobj.h>
#include <fstream>
#include <codecvt>
#include <locale>

namespace fs = std::filesystem;

// 将 UTF-8 转换为 UTF-16 (Windows wchar_t)
std::wstring utf8ToUtf16(const std::string& utf8Str) {
    int size_needed = MultiByteToWideChar(CP_UTF8, 0, utf8Str.c_str(), -1, nullptr, 0);
    std::wstring wstrTo(size_needed, 0);
    MultiByteToWideChar(CP_UTF8, 0, utf8Str.c_str(), -1, &wstrTo[0], size_needed);
    return wstrTo;
}

// 将 UTF-16 转换为 UTF-8
std::string utf16ToUtf8(const std::wstring& utf16Str) {
    int size_needed = WideCharToMultiByte(CP_UTF8, 0, utf16Str.c_str(), -1, nullptr, 0, nullptr, nullptr);
    std::string strTo(size_needed, 0);
    WideCharToMultiByte(CP_UTF8, 0, utf16Str.c_str(), -1, &strTo[0], size_needed, nullptr, nullptr);
    return strTo;
}

// 检查文本是否包含中文字符
bool containsChinese(const std::string& text) {
    // 将 UTF-8 转换为 UTF-16 (Windows wchar_t)
    std::wstring wstr = utf8ToUtf16(text);
    
    // 检查是否包含中文字符
    for (wchar_t wc : wstr) {
        if (wc >= 0x4e00 && wc <= 0x9fff) {
            return true;
        }
    }
    return false;
}

// 创建黑色背景带文字的图像
cv::Mat createBlackImageWithText(const std::string& text, int width = 1920, int height = 1080) {
    cv::Mat img = cv::Mat::zeros(height, width, CV_8UC3);
    cv::Point textOrg(width / 2, height / 2);
    int fontFace = cv::FONT_HERSHEY_SIMPLEX;
    double fontScale = 2.0;
    int thickness = 3;
    cv::Scalar textColor(255, 255, 255); // 白色文字
    
    // 获取文本大小来居中显示
    int baseline = 0;
    cv::Size textSize = cv::getTextSize(text, fontFace, fontScale, thickness, &baseline);
    textOrg.x = (width - textSize.width) / 2;
    textOrg.y = (height + textSize.height) / 2;
    
    cv::putText(img, text, textOrg, fontFace, fontScale, textColor, thickness);
    return img;
}

// 执行系统命令
int executeCommand(const std::string& command) {
    std::cout << "Executing command: " << command << std::endl;
    return system(command.c_str());
}

// 选择文件夹对话框
std::string selectFolder() {
    BROWSEINFOA bi = { 0 };
    bi.lpszTitle = "Select folder containing video files";
    bi.ulFlags = BIF_RETURNONLYFSDIRS | BIF_NEWDIALOGSTYLE;
    
    LPITEMIDLIST pidl = SHBrowseForFolderA(&bi);
    if (pidl == NULL) {
        return "";
    }
    
    char path[MAX_PATH];
    if (SHGetPathFromIDListA(pidl, path)) {
        CoTaskMemFree(pidl);
        return std::string(path);
    }
    
    CoTaskMemFree(pidl);
    return "";
}

int main() {
    // 初始化 COM 库 (用于 Windows 对话框)
    CoInitializeEx(NULL, COINIT_APARTMENTTHREADED | COINIT_DISABLE_OLE1DDE);
    
    // 选择文件夹
    std::string folderPath = selectFolder();
    if (folderPath.empty()) {
        std::cout << "No folder selected." << std::endl;
        CoUninitialize();
        return 1;
    }
    
    std::cout << "Selected folder: " << folderPath << std::endl;
    
    // 获取文件夹名称
    std::string folderName = fs::path(folderPath).filename().string();
    std::cout << "Folder name: " << folderName << std::endl;
    
    // 创建临时文件夹
    std::string tempFolder = folderPath + "\\temp";
    if (fs::exists(tempFolder)) {
        for (const auto& entry : fs::directory_iterator(tempFolder)) {
            fs::remove_all(entry.path());
        }
    } else {
        fs::create_directory(tempFolder);
    }
    
    // 获取所有视频文件
    std::vector<std::string> mainVideos;
    std::vector<std::string> otherVideos;
    
    for (const auto& entry : fs::directory_iterator(folderPath)) {
        if (entry.is_regular_file()) {
            std::string extension = entry.path().extension().string();
            std::transform(extension.begin(), extension.end(), extension.begin(), ::tolower);
            
            if (extension == ".mp4" || extension == ".avi" || extension == ".mov" || extension == ".mkv") {
                std::string filename = entry.path().filename().string();
                std::string lowerFilename = filename;
                std::transform(lowerFilename.begin(), lowerFilename.end(), lowerFilename.begin(), ::tolower);
                
                if (lowerFilename.find("main") != std::string::npos) {
                    mainVideos.push_back(entry.path().string());
                } else {
                    otherVideos.push_back(entry.path().string());
                }
            }
        }
    }
    
    // 按名称排序
    std::sort(mainVideos.begin(), mainVideos.end());
    std::sort(otherVideos.begin(), otherVideos.end());
    
    // 合并主视频和其他视频
    std::vector<std::string> allVideos;
    allVideos.insert(allVideos.end(), mainVideos.begin(), mainVideos.end());
    allVideos.insert(allVideos.end(), otherVideos.begin(), otherVideos.end());
    
    if (allVideos.empty()) {
        std::cout << "No video files found in the folder!" << std::endl;
        CoUninitialize();
        return 1;
    }
    
    // 收集所有需要合并的视频
    std::vector<std::string> allVideosToMerge;
    int videoIndex = 0;
    
    for (const auto& videoPath : allVideos) {
        fs::path path = fs::path(videoPath);
        std::string videoFileName = path.filename().string();
        std::cout << "Processing video: " << videoFileName << std::endl;
        
        // 创建黑色背景带文字的图像
        cv::Mat titleImage = createBlackImageWithText(videoFileName);
        std::string titleImagePath = tempFolder + "\\title_" + std::to_string(videoIndex) + ".jpg";
        cv::imwrite(titleImagePath, titleImage);
        
        // 创建1秒图片视频
        std::string titleVideoPath = tempFolder + "\\title_" + std::to_string(videoIndex) + ".mp4";
        std::string ffmpegCmd = "ffmpeg -y -loop 1 -i \"" + titleImagePath + 
                               "\" -c:v libx264 -t 1 -pix_fmt yuv420p -r 30 \"" + titleVideoPath + "\"";
        if (executeCommand(ffmpegCmd) != 0) {
            std::cout << "Failed to create title video!" << std::endl;
            continue;
        }
        
        // 处理原始视频 (缩放到1080p并保持宽高比)
        std::string scaledVideoPath = tempFolder + "\\scaled_" + std::to_string(videoIndex) + ".mp4";
        ffmpegCmd = "ffmpeg -y -i \"" + videoPath + 
                   "\" -vf \"scale=1920:1080:force_original_aspect_ratio=decrease,pad=1920:1080:(ow-iw)/2:(oh-ih)/2:black\" " +
                   "-c:v libx264 -c:a aac \"" + scaledVideoPath + "\"";
        if (executeCommand(ffmpegCmd) != 0) {
            std::cout << "Failed to scale video!" << std::endl;
            continue;
        }
        
        // 添加标题视频和缩放后的视频到合并列表
        allVideosToMerge.push_back(titleVideoPath);
        allVideosToMerge.push_back(scaledVideoPath);
        
        videoIndex++;
    }
    
    // 检查 disclaimer 视频是否存在
    std::string disclaimerPath = "F:\\V_of_paper\\disclaimer_video.mp4";
    if (!fs::exists(disclaimerPath)) {
        std::cout << "Disclaimer video does not exist: " << disclaimerPath << std::endl;
        
        // 创建临时免责声明视频
        cv::Mat disclaimerImage = createBlackImageWithText("DISCLAIMER\nFor educational purposes only");
        std::string disclaimerImagePath = tempFolder + "\\disclaimer.jpg";
        cv::imwrite(disclaimerImagePath, disclaimerImage);
        
        disclaimerPath = tempFolder + "\\disclaimer_video.mp4";
        std::string ffmpegCmd = "ffmpeg -y -loop 1 -i \"" + disclaimerImagePath + 
                              "\" -c:v libx264 -t 3 -pix_fmt yuv420p -r 30 \"" + disclaimerPath + "\"";
        executeCommand(ffmpegCmd);
    }
    
    // 将 disclaimer 视频添加到合并列表
    allVideosToMerge.push_back(disclaimerPath);
    
    // 创建最终合并文件列表
    std::string finalConcatListPath = tempFolder + "\\final_concat.txt";
    std::ofstream finalConcatList(finalConcatListPath);
    for (const auto& video : allVideosToMerge) {
        finalConcatList << "file '" << video << "'\n";
    }
    finalConcatList.close();
    
    // 最终一次性合并所有视频
    std::string outputPath = "F:\\V_of_paper\\" + folderName + "merge.mp4";
    std::string finalCommand = "ffmpeg -y -f concat -safe 0 -i \"" + finalConcatListPath + 
                              "\" -c:v libx264 -crf 23 -preset medium -c:a aac \"" + outputPath + "\"";
    
    if (executeCommand(finalCommand) == 0) {
        std::cout << "Video merged successfully! Output file: " << outputPath << std::endl;
    } else {
        std::cout << "Final video merge failed!" << std::endl;
    }
    
    CoUninitialize();
    
    std::cout << "Processing completed. Press any key to exit..." << std::endl;
    std::cin.get();
    
    return 0;
}

File: src/merge_video.py
import re
import os
import subprocess
import cv2
import numpy as np
from tkinter import Tk, filedialog
from concurrent.futures import ThreadPoolExecutor
from moviepy import VideoFileClip, concatenate_videoclips
output_path = f'F:/文献的视频资料/'

def contains_chinese(text):
    """检查文本是否包含中文字符"""
    chinese_pattern = re.compile('[\u4e00-\u9fff]')
    return bool(chinese_pattern.search(text))

def create_black_image_with_text(text, width=1920, height=1080):
    img = np.zeros((height, width, 3), dtype=np.uint8)
    font = cv2.FONT_HERSHEY_SIMPLEX
    font_scale = 2
    font_thickness = 3
    line_spacing = 20  # 每行文字之间的间距（像素）

    # 将文本按照换行符进行分割
    lines = text.split('\n')

    # 处理每一行，按宽度拆分多行
    wrapped_lines = []
    for line in lines:
        current_line = ""
        for word in line.split():
            test_line = current_line + (" " if current_line else "") + word
            text_size = cv2.getTextSize(test_line, font, font_scale, font_thickness)[0]
            if text_size[0] <= width - 40:  # 留边距
                current_line = test_line
            else:
                wrapped_lines.append(current_line)
                current_line = word
        if current_line:  # 加入最后未保存的一行
            wrapped_lines.append(current_line)

    # 计算文字开始绘制的初始 y 坐标，使文字垂直居中
    total_text_height = len(wrapped_lines) * (cv2.getTextSize("Test", font, font_scale, font_thickness)[0][1] + line_spacing)
    start_y = (height - total_text_height) // 2

    # 绘制每行文字
    y_offset = start_y
    for line in wrapped_lines:
        text_size = cv2.getTextSize(line, font, font_scale, font_thickness)[0]
        text_x = (width - text_size[0]) // 2
        text_y = y_offset + text_size[1]
        cv2.putText(img, line, (text_x, text_y), font, font_scale, (255, 255, 255), font_thickness)
        y_offset += text_size[1] + line_spacing

    return img


def process_video_file(i, video_file, temp_folder, create_video):
    final_videos = []
    try:
        if create_video:
            # Step 1: Create black image and add text
            image_with_text = create_black_image_with_text(video_file)
            temp_image_file = os.path.join(temp_folder, f"temp_image_{i}.jpg")
            cv2.imwrite(temp_image_file, image_with_text)
            temp_image_video_file = os.path.join(temp_folder, f"temp_image_video_{i}.mp4")
            
            command_image_video = [
                "ffmpeg", "-loop", "1", "-framerate", "1", "-t", "1",
                "-i", temp_image_file,
                "-f", "lavfi", "-t", "1", "-i", "anullsrc=r=44100:cl=stereo",
                "-c:v", "libx264", "-pix_fmt", "yuv420p", "-r", "30",
                temp_image_video_file
            ]
            subprocess.run(command_image_video, check=True)
            final_videos.append(temp_image_video_file)

        # Step 2: Scale video to 1080p and change framerate
        scaled_file = os.path.join(temp_folder, f"temp_scaled_{i}.mp4")
        command_scale = [
            "ffmpeg", "-i", video_file,
            "-vf", "scale=w=1920:h=1080:force_original_aspect_ratio=decrease,pad=1920:1080:(1920-iw)/2:(1080-ih)/2",
            "-r", "30",
            "-c:v", "libx264",
            "-preset", "fast",
            "-crf", "23",
            "-c:a", "aac",
            scaled_file
        ]
        subprocess.run(command_scale, check=True)
        final_videos.append(scaled_file)
        
    except subprocess.CalledProcessError as e:
        print(f"处理视频 {video_file} 时出错: {e}")

    return final_videos

def merge_videos_in_current_path(folder_name,output_path):
    video_files = [
        f for f in sorted(os.listdir('.'))
        if os.path.isfile(f)
        and f.lower().endswith(('.mp4', '.avi', '.mov', '.mkv'))
        and not f.endswith('-merge.mp4')
    ]
    # 优先将包含 "main" 的文件提取出来
    videos_with_main = [f for f in video_files if 'main' in f.lower()]
    videos_without_main = [f for f in video_files if 'main' not in f.lower()]
    # 合并列表，优先显示包含 "main" 的视频
    video_files = videos_with_main + videos_without_main
    if not video_files:
        print("当前路径下没有找到有效的视频文件。")
        return
    temp_folder = os.path.join('.', 'temp_folder')
    if os.path.exists(temp_folder):
            for temp_file in os.listdir(temp_folder):
                os.remove(os.path.join(temp_folder, temp_file))
            os.rmdir(temp_folder)
    if not os.path.exists(temp_folder):
        os.makedirs(temp_folder)
    output_file = os.path.join(output_path, f"{folder_name}-merge.mp4")
    # If output file exists, delete it
    if os.path.exists(output_file):
        os.remove(output_file)

    # Use an existing local video as the disclaimer video
    existing_disclaimer_video = "F:/文献的视频资料/已发/disclaimer_video.mp4"
    if not os.path.exists(existing_disclaimer_video):
        print(f"指定的免责声明视频文件不存在: {existing_disclaimer_video}")
            # Create a disclaimer image with the required text
        disclaimer_text = f'For educational purposes only, not for commercial use.\nIf there is any infringement, please contact for removal.'
        disclaimer_image = create_black_image_with_text(disclaimer_text)
        disclaimer_image_file = os.path.join(temp_folder, "disclaimer_image.jpg")
        cv2.imwrite(disclaimer_image_file, disclaimer_image)
        temp_disclaimer_video_file = os.path.join(temp_folder, "disclaimer_image_video.mp4")

        # Generate the disclaimer video (1 second duration)
        command_disclaimer_video = [
            "ffmpeg", "-loop", "1", "-framerate", "1", "-t", "3",  # 3 seconds for visibility
            "-i", disclaimer_image_file,
            "-f", "lavfi", "-t", "3", "-i", "anullsrc=r=44100:cl=stereo",
            "-c:v", "libx264", "-pix_fmt", "yuv420p", "-r", "30",
            temp_disclaimer_video_file
        ]
        subprocess.run(command_disclaimer_video, check=True)
    else:
        temp_disclaimer_video_file = existing_disclaimer_video
    # Create a thread pool to parallelize the processing of videos
    with ThreadPoolExecutor() as executor:
        final_videos = []
        
        for i, video_file in enumerate(video_files):
            create_video = not contains_chinese(video_file)  # Skip creating video for Chinese text files
            future = executor.submit(process_video_file, i, video_file, temp_folder, create_video)
            final_videos.extend(future.result())
        
        # Step 3: Merge all videos including disclaimer image video at the end
        final_videos.append(temp_disclaimer_video_file)  # Add the disclaimer video at the end

        temp_list_file = os.path.join(temp_folder, "temp_file_list.txt")
        with open(temp_list_file, "w") as f:
            for video_file in final_videos:
                f.write(f"file '{video_file}'\n")
        # 加载视频文件
        clips = [VideoFileClip(video_file) for video_file in final_videos]
        # 合并视频
        final_clip = concatenate_videoclips(clips, method="compose")
        # 保存合并后的视频
        final_clip.write_videofile(output_file, codec="libx264", fps=30)

        print(f"视频已成功合并，并保存为: {output_file}")
        #Clean up
        if os.path.exists(temp_folder):
            for temp_file in os.listdir(temp_folder):
                os.remove(os.path.join(temp_folder, temp_file))
            os.rmdir(temp_folder)

# Main entry
if __name__ == "__main__":
    Tk().withdraw()  # Hide the root window
    folder_path = filedialog.askdirectory(title="选择包含视频文件的文件夹")
    if not folder_path:
        print("未选择文件夹。")
    else:
        os.chdir(folder_path)
        folder_name = os.path.basename(folder_path)
        merge_videos_in_current_path(folder_name,output_path)
    os.system('pause')